tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -226,6 +226,295 @@ static bool HasCanonicalDimensionNumbers(
+// rhs:[d2,d3], out:[d0,d1,d3]}.
+class EinsumToLinalgConverter : public OpConversionPattern<mhlo::EinsumOp> {
+ public:
+  using OpConversionPattern<mhlo::EinsumOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mhlo::EinsumOp op, ArrayRef<Value> args,
+      ConversionPatternRewriter& rewriter) const final {
+    mhlo::EinsumOp::Adaptor adaptor(args);
+    auto get_rank = [](Value v) {
+      return v.getType().cast<ShapedType>().getRank();
+    };
+    auto einsum_config = op.einsum_config();
+
+    // With the assumption of binary input operand and single output
+    // get the inputs and output operands' indices.
+    // einsum_config = "lhs_loop,rhs_loop->out_loop"
+    std::size_t pos_arrow = einsum_config.find(arrow_);
+    std::size_t pos_comma = einsum_config.find(comma_);
+
+    StringRef lhs_loop = einsum_config.substr(0, pos_comma);
+    StringRef rhs_loop = einsum_config.substr(
+        pos_comma + comma_.size(), pos_arrow - (pos_comma + comma_.size()));
+    StringRef out_loop = einsum_config.substr(pos_arrow + arrow_.size());
+
+    // Check for Invalid Configs.
+    // 1.Check that there is only maximum 2 inputs
+    // 2.Check that there is only maximum 1 output
+    // 3.Check that there is 1 arrow_
+    if (rhs_loop.find(comma_) != std::string::npos ||
+        out_loop.find(comma_) != std::string::npos ||
+        out_loop.find(arrow_) != std::string::npos) {
+      return rewriter.notifyMatchFailure(op, "Invalid einsum config!");
+    }
+
+    // Find result type, if on tensors.
+    auto result_ty = this->typeConverter
+                         ->convertType(GetHloOpResultType</*isLHLO=*/false>(op))
+                         .dyn_cast<RankedTensorType>();
+
+    // Check result type compatibility.
+    if (!result_ty || !(result_ty.getElementType().isSignlessIntOrFloat())) {
+      return rewriter.notifyMatchFailure(op, "Invalid result type");
+    }
+
+    // Convert the representation to vector<string>.
+    SmallVector<std::string> lhs_ein =
+        GetEinsumConfigAsVector(lhs_loop, get_rank(adaptor.lhs()));
+    SmallVector<std::string> rhs_ein =
+        GetEinsumConfigAsVector(rhs_loop, get_rank(adaptor.rhs()));
+    SmallVector<std::string> out_ein =
+        GetEinsumConfigAsVector(out_loop, result_ty.getRank());
+    SmallVector<SmallVector<std::string>> all_loop_dim = {lhs_ein, rhs_ein,
+                                                          out_ein};
+
+    if (!CheckBatchHasEqualRank(lhs_ein.size(), lhs_loop, rhs_ein.size(),
+                                rhs_loop, out_ein.size(), out_loop)) {
+      return rewriter.notifyMatchFailure(
+          op, "Invalid elipsis('...') within einsum config!");
+    }
+
+    // Find all unique indices in the input and output.
+    llvm::SmallSetVector<StringRef, 4> input_ind;
+    llvm::SmallSetVector<StringRef, 4> output_ind;
+    for (const auto& loops_it : llvm::enumerate(all_loop_dim)) {
```

Comment:
ditto

Reply:
ah, not sure if it is, maybe we can spell it out?

Reply:
Changed the inside of the loop to be a function, and calling it three times for each operand instead. so we can clear out the redundant variable "all_loop_dim" and also make the code more understandable :)
lmk what you think(whether or not we should revert back to the original loop)

Reply:
Ok, ended up removing this loop and using llvm's setvector's functionality and calling it multiple times. S.T we can get rid of redundant all_loop_dim and get better readability :)